crates/bevy_ecs/src/traversal.rs

```diff
@@ -0,0 +1,39 @@
+//! A trait for components that let you traverse the ECS.
+
+use crate::{
+    component::{Component, StorageType},
+    entity::Entity,
+};
+
+/// A component that can point to another entity, and which can be used to define a path through the ECS.
+///
+/// Traversals are used to [specify the direction] of [event propagation] in [observers]. By default,
+/// events use the [`TraverseNone`] placeholder component, which cannot actually be created or added to
+/// an entity and so never causes traversal.
+///
+/// The implementer is responsible for ensuring that `Traversal::next` cannot produce infinite loops.
```

Comment:
This isn't actually invalid is it? You can have an observer that triggers itself until it decides to stop propagation.
I'd keep it as a warning but not as a contract/responsibility.

Reply:
That's a good point. I'll add some more discussion about this.